File: LeetcodeQuestions/countBinarySubstrings-696.js
var countBinarySubstrings = function(s) {
    let counter = 0;
    if (s.length === 1) return 0;
    
    let prevBlock = 0;
    let currentBlock = 1;
    let currentChar = s[0];
    
    for (let i = 1; i < s.length; i++) {
        if (currentChar === s.charAt(i)) {
            currentBlock++;
        } else {
            counter += Math.min(currentBlock, prevBlock);
            prevBlock = currentBlock;
            currentBlock = 1;
        }
        currentChar = s.charAt(i);
    }
    
    counter += Math.min(currentBlock, prevBlock);
    return counter;
};

console.log(countBinarySubstrings("00110011"))
console.log(countBinarySubstrings("10101"))

File: LeetcodeQuestions/count_binary_substrings-696.rb
def count_binary_substrings(s)
    counter = 0
    prev_block = 0
    curr_block = 1
    
    current = s[0]
    
    (1...s.length).each do |i|
        if current == s[i]
            curr_block += 1
        else 
            counter += [curr_block, prev_block].min
            prev_block = curr_block
            curr_block = 1
        end
        current = s[i]
    end
    counter += [curr_block, prev_block].min
    counter
end

puts count_binary_substrings("00110011")
puts count_binary_substrings("10101")